### PR TITLE
Implement Zarr V3 protocol

### PR DESCRIPTION
**Pull request by grlee77**
This PR builds on #897, adding zarr v3 support ~to most of the high-level convenience functions (I haven't looked at the consolidated metadata yet).~

**EDIT**: I closed the prior v3 PRs leading up to this one as their commits are also included here along with additional bug fixes and tests. 

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [x] GitHub Actions have all passed
* [x] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by joshmoore** [comment]
cc: @d-v-b

**Comment by joshmoore** [comment]
> Also should we really have both `NestedDirectoryStoreV3` and `DirectoryStoreV3`?

My vote would be no.  

> The v3 spec has a separator attribute in the array's chunk_grid metadata so it is probably sufficient to provide only DirectoryStoreV3?

Agreed.

**Comment by grlee77** [comment]
I opened https://github.com/zarr-developers/zarr-specs/pull/135 regarding the other dtype extensions which currently have missing URLs in `meta.py` here

**Comment by grlee77** [comment]
`NestedDirectoryStoreV3` has been removed as discussed above: https://github.com/zarr-developers/zarr-python/pull/898#issuecomment-1057147337

**Review by joshmoore — COMMENTED** [review]
Looking nice, @grlee77! A number of mostly aesthetic points below after an initial read. Don't think any of them are MUSTs before merging this into a dev branch. (See https://github.com/zarr-developers/governance/issues/15)

The only other high-level comment, also aesthetic, is whether `zarr._storage.v3` (or similar) would be appropriate earlier rather than later.

File: zarr/_storage/absstore.py [hunk 1]
```diff
@@ -209,3 +205,63 @@ def getsize(self, path=None):
 
     def clear(self):
         self.rmdir()
+
+
+def rmdir_abs(store: ABSStore, path=None):
```
[the remark this answers is not in the document]
Comment (grlee77) [reply zarr/_storage/absstore.py:210]: Not sure why I did it that way. I can make this `ABSStore.rmdir` and just call it from there in `ABSStoreV3.rmdir`

File: zarr/_storage/store.py [hunk 2]
```diff
@@ -131,6 +137,165 @@ def rmdir(self, path: str = "") -> None:
         _rmdir_from_keys(self, path)
 
 
+_valid_key_characters = set(ascii_letters + digits + "/.-_")
```
[the remark this answers is not in the document]
Comment (grlee77) [reply zarr/_storage/store.py]: this has been moved into the class

File: zarr/_storage/store.py [hunk 3]
```diff
@@ -140,24 +305,94 @@ def _path_to_prefix(path: Optional[str]) -> str:
     return prefix
 
 
+# TODO: Should this return default metadata or raise an Error if zarr.json
```
[the remark this answers is not in the document]
Comment (grlee77) [reply zarr/_storage/store.py]: Sure, this now raises an error if the key is not present

File: zarr/convenience.py [hunk 4]
```diff
@@ -1118,9 +1193,29 @@ def consolidate_metadata(store: StoreLike, metadata_key=".zmetadata"):
...
+    else:
+
+        sfx = _get_hierarchy_metadata(store)['metadata_key_suffix']
+
+        def is_zarr_key(key):
+            return (key.endswith('.array' + sfx) or key.endswith('.group' + sfx) or
+                    key == 'zarr.json')
+
+        # cannot create a group without a path in v3
+        # so create /meta/root/consolidated group to store the metadata
```
[the remark this answers is not in the document]
Comment (grlee77) [reply zarr/convenience.py:1219]: I opened https://github.com/zarr-developers/zarr-specs/issues/136 so we don't forget

File: zarr/hierarchy.py [hunk 5]
```diff
@@ -1104,20 +1233,29 @@ def group(store=None, overwrite=False, chunk_store=None,
     """
 
     # handle polymorphic store arg
-    store = _normalize_store_arg(store)
+    store = _normalize_store_arg(store, zarr_version=zarr_version)
+    if zarr_version is None:
+        zarr_version = getattr(store, '_store_version', 2)
```
[the remark this answers is not in the document]
Comment (grlee77) [reply zarr/hierarchy.py]: Okay, I added `DEFAULT_ZARR_VERSION` (currently set to 2) in place of 2 in these checks.

It is intended as a constant used by developers, not something a user would set in their code to select the default version.

**Comment by joshmoore** [comment]
Anything else planned from your side, @grlee77 ?

**Comment by grlee77** [comment]
> Anything else planned from your side, @grlee77 ?

Before release we will need to decide on where we want the import path for v3 Stores to be (currently in the same file as the V2 ones). We can think about that and a possible environment variable to enable/disable v3 visibility in a followup, though.,

**Comment by joshmoore** [comment]
:+1: for handling the refactoring and v3-switch in follow ups. As discussed during the past two community calls, getting this in to the mainline. :tada: (And as promised for @grlee77: 🍪)

Clear warning for @zarr-developers/core-devs : the mainline is now in an unreleasable `v3` state. If 2.11.x needs to go out, a dedicated `v2` branch will need to be forked from a point before this merge.

See https://github.com/zarr-developers/governance/issues/15 for some background.

**Pull request merged**